pkg/kubelet/dockertools/kube_docker_client.go

```diff
@@ -0,0 +1,471 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dockertools
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"strconv"
+	"time"
+
+	"github.com/docker/docker/pkg/stdcopy"
+	dockerapi "github.com/docker/engine-api/client"
+	dockertypes "github.com/docker/engine-api/types"
+	dockercontainer "github.com/docker/engine-api/types/container"
+	dockerfilters "github.com/docker/engine-api/types/filters"
+	docker "github.com/fsouza/go-dockerclient"
+	"golang.org/x/net/context"
+	"k8s.io/kubernetes/pkg/kubelet/metrics"
+)
+
+// KubeDockerClient is a wrapped layer of docker client for kubelet internal use. This layer is added to:
+//	1) Record operation metrics in each function.
+//	2) Redirect stream for exec and attach operations.
+//	3) Wrap the context in this layer to make the DockerInterface cleaner.
+//	3) Stabilize the DockerInterface. The engine-api is still under active development, the interface
+//	is not stabilized yet. However, the DockerInterface is used in many files in Kubernetes, we may
+//	not want to change the interface frequently. With this layer, we can port the engine api to the
+//	DockerInterface to avoid changing DockerInterface as far as possible.
```

Comment:
@yujuhong Yeah, this is just the first step to make the following refactoring easier, I'll add a TODO here.